### DISCUSSION TITLE
Adding `woocommerce` class to the `<body>` when using WooCommerce Blocks

### DISCUSSION THREAD

**Discussion by octaedro** [woocommerce/woocommerce#51098] [WooCommerce, Blocks, and Site Editing]
I wanted to open up a discussion around the recent fix in [PR #51023](https://github.com/woocommerce/woocommerce/pull/51023) that addressed some visual issues with the variation selector on the front end. While the fix improves the appearance of the variation selector, it also raises a broader concern about how we apply WooCommerce-related classes, particularly the `woocommerce` class, to pages containing our blocks.

Currently, we add the following classes to the page's `body`:
- **woocommerce-cart** -> to the cart page.
-  **woocommerce-checkout** -> to the checkout page.
- **woocommerce-account** -> to the account page.
- and the class **woocommerce** to the other WooCommerce pages like `product` and `shop` pages.

The other `posts` and `pages` don't have the mentioned classes above.

In the [PR comments](https://github.com/woocommerce/woocommerce/pull/51023#pullrequestreview-2270583078) (and especially in a Slack thread with @ObliviousHarmony ), we discussed the potential benefits and challenges of automatically adding the `woocommerce`, `woocommerce-account`, `woocommerce-cart` and `woocommerce-checkout` classes to the `<body>` whenever one of those blocks is present on a page or post. This approach could help maintain consistency in styling, but it also brings up some points for consideration:

1. **Impact on Third-Party Extensions:** There is a risk that third-party developers might find WooCommerce styles unexpectedly applied to their extensions, potentially leading to compatibility issues.

2. **Class Specificity Conflicts:** We also need to consider scenarios where multiple WooCommerce blocks are used on the same page (e.g., a `Single Product` block and a `Checkout` block). Determining which class should take precedence, or if both should be applied, is crucial to avoid CSS conflicts.

3. **Backward Compatibility:** Many themes and extensions rely on the existing behavior, and any changes could introduce breaking changes, particularly on pages like the Cart, Checkout, and Account pages, which use specific classes like `woocommerce-cart`, `woocommerce-checkout`, and `woocommerce-account`.

Given these considerations, it would make sense to explore the idea of adding the `woocommerce` class to the `<body>` tag for pages that include WooCommerce blocks. However, we need to ensure that this solution is robust and minimizes the risk of unintended side effects.

### Key Questions

- What criteria should we use to determine when and which classes should be applied to the `<body>` tag?

- How can we ensure backward compatibility while implementing this change?

- Are there any specific testing strategies or edge cases we should consider before rolling out such a change?

Looking forward to hearing your thoughts and feedback on this.

**Comment by Aljullu** [comment]
Thanks for starting this discussion, @octaedro! It's a very interesting topic which I don't think has a perfect solution.

My main concern with adding `.woocommerce` classes to the body element of a page containing a WC block is that it might cause unexpected styling side-effects. For example, a theme might be doing `.woocommerce a { color: uppercase; }`, intentionally only targeting WC templates, but once a block is added to a random page, all links, even the ones not related to WooCommerce, would end up being uppercase.

I personally feel it's safer to add the `.woocommerce` class only to the block wrapper element. Ideally, I would only do that in cases where we intentionally want classic theme styles to leak into our blocks (https://github.com/woocommerce/woocommerce/pull/51023 is a good example of this, IMO). But for blocks where we don't need to support classic themes CSS, I would avoid adding the `.woocommerce` class (I'm thinking of the filter blocks, for example, which have a different markup than the filter widgets).

In the end, we should encourage styling WooCommerce blocks via global styles (using the Site Editor UI or the `theme.json` file) and not with CSS.
